src/main/daml/Daml/Finance/Settlement/Instruction.daml

```diff
@@ -43,10 +46,13 @@ template Instruction
     signatory requestors, signed
     observer step.sender, step.receiver, settler, Disclosure.flattenObservers observers
 
-    key (requestors, id) : (Parties, Id)
-    maintainer fst key
+    key (requestors, tradeId, id) : (Parties, Id, Id)
+    maintainer fst3 key
 
-    let messageSuffix = " / instruction id = " <> show id
+    let
+      messageSuffix = " / instruction id = " <> show id
+      discloseAccount account discloser = Account.exerciseInterfaceByKey @Disclosure.I account discloser Disclosure.AddObservers with disclosers = singleton discloser; observersToAdd = (show tradeId, singleton settler)
+      discloseTransferable : ContractId Transferable.I -> Party -> Update (ContractId Transferable.I) = \transferableCid discloser -> coerceContractId <$> exercise (toInterfaceContractId @Disclosure.I transferableCid) Disclosure.AddObservers with disclosers = singleton discloser; observersToAdd = (show tradeId, singleton settler)
```

Comment:
Perhaps you should make these top level functions? It would mean they won't get recalculated each time the template is modified and if they are useful outside of this module, they can be called by other choices.

Also, `discloseTransferable` is written using lambda syntax which I believe we are to avoid (even thought personally I prefer this syntax 😕).

Reply:
I factored it out into utility functions.